src/extension.ts

```diff
@@ -1,95 +1,119 @@
-			}
-
-			if (stderr) {
-				outputChannel.appendLine(stderr);
-			}
-
-			outputChannel.show();
-		}
-	});
-};
-
-const checkConfiguration = () => {
-	const config = getConfig();
-
-	if (!config.path) {
-		vscode.window.showInformationMessage(messages.ConEmuPathNotConfigured, messages.OpenSettings).then(openSettingsCallback);
-		return false;
-	}
-
-	if (!fs.existsSync(config.path)) {
-		vscode.window.showInformationMessage(messages.ConEmuPathInvalid, messages.OpenSettings).then(openSettingsCallback);
-		return false;
-	}
-
-	return true;
-};
-
-const getConfig = () => vscode.workspace.getConfiguration("ConEmu") as any as IConfig;
-
-const openSettingsCallback = (btn) => {
-	if (btn === messages.OpenSettings) {
-		vscode.commands.executeCommand("workbench.action.openGlobalSettings");
-	}
-};
-interface IConfig {
-	path: string;
-	reuseInstance: boolean;
-	showMessageInOutputPanel: boolean;
-}
-
-const messages = {
-	WindowsOnly: "This extension works only on Windows, sorry",
-	ShowInfo: "Show Info",
-	ReadmeUrl: "https://github.com/ipatalas/vscode-conemu/blob/master/README.md",
-	ConEmuPathNotConfigured: "ConEmu path is not configured. Set proper path in ConEmu.path setting",
-	OpenSettings: "Open Settings",
-	ConEmuPathInvalid: "ConEmu path is invalid, please correct it."
-};
+'use strict';
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as child from 'child_process';
+
+// tslint:disable-next-line:no-var-requires
+const pkg = require('../../package.json');
+
+const isCompatiblePlatform = process.platform === 'win32';
+
+export function activate(context: vscode.ExtensionContext) {
+	if (!isCompatiblePlatform) {
+		vscode.window.showInformationMessage(messages.WindowsOnly, messages.ShowInfo).then(btn => {
+			if (btn === messages.ShowInfo) {
+				child.exec(`start ${messages.ReadmeUrl}`).unref();
+			}
+		});
+	}
+
+	context.subscriptions.push(vscode.commands.registerCommand('vscode.conemu', (uri?: vscode.Uri) => {
+		// tslint:disable-next-line:curly
+		if (!isCompatiblePlatform || !checkConfiguration()) return;
+
+		if (uri && uri.scheme && uri.scheme !== "untitled") {
+			runConEmu(path.dirname(uri.fsPath));
+		} else if (vscode.window.activeTextEditor && !vscode.window.activeTextEditor.document.isUntitled) {
+			runConEmu(path.dirname(vscode.window.activeTextEditor.document.uri.fsPath));
+		} else if (vscode.workspace.rootPath) {
+			runConEmu(vscode.workspace.rootPath);
+		}
+	}));
+}
+
+const runConEmu = (path: string) => {
+	const config = getConfig();
+	const reuseInstanceArg = config.reuseInstance ? "-Single" : "-NoSingle";
+	const showMessage = config.showMessageInOutputPanel;
+
+	const quote = (p: string) => p.includes(" ") ? `"${p}"` : p;
+
+	if (config.customRunOption) {
+		child.exec(`${quote(config.path)} -dir ${quote(path)} ${reuseInstanceArg} -run ${config.customRunOption}`, (error: Error, _stdout: string, stderr: string) => {
+			if (showMessage && (error || stderr)) {

```

Comment:
Error handling is copy-pasted now. It should be better to just make the command itself conditional (depending on `customRunOption`) and leave the rest untouched. It will be easier to maintain.

Reply:
I'm not sure I understand the request because I altered the logic in my previous update to just depend on `customRunOption`

Am I missing another reference you're making that I'm not seeing/following?

Reply:
Please notice that the error handling part of `child.exec` is just copy-pasted twice which makes it more error prone and harder to maintain in the future.
There are two blocks of code which do the same:
```TS
if (showMessage && (error || stderr)) {
	const outputChannel = vscode.window.createOutputChannel(pkg.displayName);

	if (error) {
		outputChannel.appendLine(error.message);
	}

	if (stderr) {
		outputChannel.appendLine(stderr);
	}

	outputChannel.show();
}
```

This can be simplified. What I meant is to make the command itself conditional so that depending on `customRunOption` either the `-run` parameter is added to the command or not.

Then it becomes something like that:
```TS
let command = 'command without -run parameter';
if (config.customRunOption) {
  command += `-run ${config.customRunOption}`
}

child.exec(command, err => {
  // all error handling here in one place
});
```